tests/defer_regress/tests.py

```diff
@@ -309,6 +309,27 @@ def test_only_reverse_many_to_many_ignored(self):
         with self.assertNumQueries(1):
             self.assertEqual(Item.objects.only("request").get(), item)
 
+    def test_self_referential_one_to_one(self):
+        first = Item.objects.create(name="first", value=1)
+        second = Item.objects.create(name="second", value=2, source=first)
+        with self.assertNumQueries(1):
+            deferred_first, deferred_second = (
+                Item.objects.select_related("source", "destination")
+                .only("name", "source__name", "destination__value")
+                .order_by("pk")
+            )
```

Comment:
To give you an idea of how this was broken in >4.2 @OdinsPlasmaRifle

This is the query that was performed on Django 4.1. Notice how the select mask is a union of all the references to the same model

```sql
SELECT 
  "defer_regress_item"."id", 
  "defer_regress_item"."name", 
  "defer_regress_item"."value", 
  "defer_regress_item"."source_id", 
  T2."id", 
  T2."name", 
  T2."value", 
  T2."source_id", 
  T3."id", 
  T3."name", 
  T3."value", 
  T3."source_id" 
FROM 
  "defer_regress_item" 
  LEFT OUTER JOIN "defer_regress_item" T2 ON (
    "defer_regress_item"."source_id" = T2."id"
  ) 
  LEFT OUTER JOIN "defer_regress_item" T3 ON (
    "defer_regress_item"."id" = T3."source_id"
  ) 
ORDER BY 
  "defer_regress_item"."id" ASC
```

This now doing the proper thing

```sql
SELECT 
  "defer_regress_item"."id", 
  "defer_regress_item"."name", 
  "defer_regress_item"."source_id", 
  T2."id", 
  T2."name", 
  T3."id", 
  T3."value" 
FROM 
  "defer_regress_item" 
  LEFT OUTER JOIN "defer_regress_item" T2 ON (
    "defer_regress_item"."source_id" = T2."id"
  ) 
  LEFT OUTER JOIN "defer_regress_item" T3 ON (
    "defer_regress_item"."id" = T3."source_id"
  ) 
ORDER BY 
  "defer_regress_item"."id" ASC
```